### DISCUSSION TITLE
Map viz

### DISCUSSION THREAD

**Discussion by andresp-wave** [DASDAE/dascore#122] [Show and tell]
Hello dascore members ! 

I am Andres and relatively new to DAS. I would like to contribute when possible with code or comments to the DAScore project ! At the moment, I am working in a code to do a quick map for any given patch if it has latitude and longitude (Similar to what inventory.plot() does in Obspy). It is quite preliminary (and it also requires pygmt) + I know there are already some issues with the units. Any suggestions/comments are welcome. I will try to upload the code this week once I clean it more and fix an issue I have with github in my terminal.

best,
Andres Pena
![test](https://user-images.githubusercontent.com/22261180/229629151-823ee88d-97f2-4599-8f55-44a40be28a7d.png)

**Comment by d-chambers** [comment]
Hi @andresp-wave,

I think that would be a good contribution! 

I don't have an issue with taking pygmt as an optional dependency, especially now that it is available on conda-forge. You can use the [`optional_import`](https://github.com/DASDAE/dascore/blob/8ecbc2b9fac2b06f5a354902536c07b4c68f1171/dascore/utils/misc.py#L273-L306) method found in the `utils.misc` module.

Probably start by creating a new module (python file) in the viz directory. The tests for the function should then go in the tests/test_viz directory. When running visualization tests, we generally don't show the image but just ensure the code will run. Probably the best visualization example to look at is `viz/waterfall.py` and `tests/viz/test_waterfall.py`. Also, use the  [import or skip pytest mark](https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest-importorskip) so the new tests will only run if pygmt is installed rather than failing the test suite if it isn't. 

[The developer docs](https://dascore.org/contributing/contributing.html) should be helpful (let me know if you find any parts confusing). I am also happy to help.

**Comment by OUCyf** [comment]
Hey, I am Fu, a Rice student who worked with Jonathan, recently created a tiny package for DAS geometry using gps interpolation, and I have successfully tested it in some cases. I think it can also be used for the dascore. The main features are:
- It's a 3-d (lat, lon, altitude) interpolation function, that can deal with the geometry of any shape, including but not limited to the borehole, ocean cases with changes in seafloor topography
- Provide the forward extension and backward extension of interpolation, when faced with start and endpoints.
- Define Sag Ratio (SR) to evaluate the average interpolation error: (actual spacing - nominal spacing) / nominal spacing. (According to Prof. Zefeng's code)
- The input file needs lat/lon/altitude/channel info, sometimes we just know the position for a tracked point and don't know the channel number, just set it to -999 in the input file.
- Provide a highly customized visualization function, which uses Mapbox as the backend and can easily plot a dozen kinds of maps, street view, terrain, cartoon maps, etc. 

But one downside is that you need a token of Mapbox, and if you use it for more than a limited time, it will charge your money. For this I bind the credit card, so the token cannot be shared. And Jonathan told me Mapbox has an academic version, so maybe is possible to implement it in dascore.

There are some examples:

![image](https://github.com/DASDAE/dascore/assets/25765271/578c01c5-85a1-4a99-a4c0-db1a126aa170)
![image](https://github.com/DASDAE/dascore/assets/25765271/22c35bbc-b285-4d44-bf7a-cb569b6b2b12)
![image](https://github.com/DASDAE/dascore/assets/25765271/faf4eaee-15dc-4087-8b9e-7be2a78c4341)

**Comment by OUCyf** [comment]
I am looking for a possible implementation for the `dascore`. Maybe I will publish this tiny package in `pypi` and `conda-forge`, then `dascore` can import this package. BTW I just stumbled upon this discussion in the middle of the night to express my opinions, maybe I can't have time to commit this code recently, but hope it is a useful discussion

**Comment by OUCyf** [comment]
And I think `mapbox` is better than `pygmt` for das geometry, especially when we have some cables in the cities, we need to recognize the scene around the roads.

**Comment by OUCyf** [comment]
The following is a demo showing how to use this tiny package.

```python
# %% 2.
import plot_map
import das_interp
import numpy as np

points = np.loadtxt('known_points.txt') # 2-d np.array: Latitude, Longitude, Altitude, Channel
spacing_interval = 2.45  # unit: m
channel_start = -10
channel_end = 104

results, known_index = das_interp.interp(points, 
                             channel_start, 
                             channel_end, 
                             spacing_interval)

np.savetxt('results.txt', results, fmt='%.7f')

plot_map.set_mapboxtoken('xxxxxxx')

plot_map.set_imgsavepath('./')


das_interp.plot(points, 
     results, 
     known_index,

     # subplot
     plot_ax_right_flag=True,
     ax_right_size=0.7,
     plot_ax_top_flag=True,
     ax_top_size=0.7,
    
     # map
     plot_map_flag=True,
     map_style=3, # 3
     compasssize=2,
     compass_accuracy=2,
     compass_textsize=4,
     compass_unit='m',
     rect=[0.12,0.05], # [left, bottom]

     # colorbar
     colorbar_size=0.2,
     colorbar_ticks_labelsize=6,
     colorbar_label_fontsize=8,

     # legend
     legend_fontsize=6,

     # save
     save_flag=True,
     save_path='./interp_map.pdf',

     # plot sag
     plot_sag_flag=True,
     sag_linewidth=6,
     sag_cmap=plt.colormaps["bwr"], 

     # plot results line
     plot_results_line_flag=True,
     results_linewidth=1, 
     results_line_color='gray',

     # plot results dot
     plot_results_dot_flag=True,
     results_markersize=1,
     results_dot_color='orange',

     # plot results label
     plot_results_label_flag=True, 
     results_label_gap=10, 

     # plot points dot
     plot_points_dot_flag=True,
     points_markersize=3,
     points_dot_color='k',

     # plot points label
     plot_points_label_flag=True, 
     points_label_gap=5, 

     # fontsize of results and points label 
     results_points_label_fontsize=6, 
    
     # fontsize of ticks and axis label for all subplots
     ticks_labelsize=6,
     axis_label_fontsize=8,

     # ticks format for all subplots
     ticks_format_latlon='%0.4f',  # '%0.4f'
     ticks_format_altitude='%0.1f',  # '%0.1f'
    
     # ticks number for all subplots
     ticks_lon_num=5,
     ticks_lat_num=5,
     ticks_altitude_num=3,
     
     # ticks rotation for all subplots
     ticks_lon_rotation=30,
     ticks_lat_rotation=0,
     ticks_altitude_rotation=0,
    )

```

**Comment by d-chambers** [comment]
Hey @OUCyf,

Nice work on the plots, they look good and seem to be simple to create. The level of detail and style options are impressive. 

Two thoughts here:

1. Mapbox, although a very nice looking package, would probably be too much of a dependency for DASCore (especially if it requires registration/a credit-card). DASCore, by design, should have very limited set of dependencies. However, DASDAE aims to foster an ecosystem of DAS packages and `das_interp`  would fit very nicely in that category. The only requirement is that DASCore data structures should be supported (e.g., you should be able to pass a `Patch` or the not yet created `Inventory` to your plotting function). We are going to create a DASDAE website and package registry in the coming months/weeks which will have more details.

2. The interpolation logic could be useful to DASCore. We are working on an `Inventory`-type object (cc @eileenrmartin) so interpolating the tap test locations to fill in values for all channels in a patch and adding corresponding coordinates is something we have planned, but we need to implement the Inventory stuff first.

**Comment by OUCyf** [comment]
Yeah, I agree with your points, a package should have very limited dependencies. And interpolation logic could be useful to DASCore, and it can be added in. As for details of the `inventory` type, just let me know when you finish the implementation of the inventory stuff.
